beam-postgres/Database/Beam/Postgres/PgSpecific.hs

```diff
@@ -410,6 +413,147 @@ isSubsetOf_ (QExpr needles) (QExpr haystack) =
+    [ fromPgExpression <$> el
+    , pure (emit ", ")
+    , fromPgExpression <$> arr
+    ])
+
+-- | Postgres array_remove(value) function.
+--
+-- Removes all elements equal to the given value from the array.
+-- Comparisons use @IS NOT DISTINCT FROM@ semantics, so this can remove NULLs.
+--
+-- Notes:
+-- - The array must be one-dimensional.
+-- - Returns NULL only if the array is NULL; if the value is not present,
+--   the original array is returned unchanged.
+arrayRemove_
+  :: QGenExpr ctxt Postgres s (V.Vector a)
+  -> QGenExpr ctxt Postgres s a
+  -> QGenExpr ctxt Postgres s (V.Vector a)
+arrayRemove_ (QExpr arr) (QExpr el) =
+  QExpr (PgExpressionSyntax . mappend (emit "array_remove") . pgParens . mconcat <$> sequenceA
+    [ fromPgExpression <$> arr
+    , pure (emit ", ")
+    , fromPgExpression <$> el
+    ])
+
+-- | Postgres array_replace(array, from, to) function.
+--
+-- Replaces each element equal to the second argument with the third.
+--
+-- Notes:
+-- - Comparisons use IS NOT DISTINCT FROM semantics; can replace NULLs.
+-- - The array must be one-dimensional.
+--
+-- Example:
+--
+-- @
+-- select_ $ pure $ arrayReplace_ (val_ $ V.fromList [1::Int32,2,5,4]) (val_ 5) (val_ 3)
+-- -- => {1,2,3,4}
+-- @
+arrayReplace_
+  :: QGenExpr ctxt Postgres s (V.Vector a)
+  -> QGenExpr ctxt Postgres s a
+  -> QGenExpr ctxt Postgres s a
+  -> QGenExpr ctxt Postgres s (V.Vector a)
+arrayReplace_ (QExpr arr) (QExpr fromVal) (QExpr toVal) =
+  QExpr (PgExpressionSyntax . mappend (emit "array_replace") . pgParens . mconcat <$> sequenceA
+    [ fromPgExpression <$> arr
+    , pure (emit ", ")
+    , fromPgExpression <$> fromVal
+    , pure (emit ", ")
+    , fromPgExpression <$> toVal
+    ])
+
+-- | Postgres array_shuffle(array) function.
+-- Randomly shuffles the first dimension.
+arrayShuffle_
+  :: QGenExpr ctxt Postgres s (V.Vector a)
+  -> QGenExpr ctxt Postgres s (V.Vector a)
+arrayShuffle_ (QExpr arr) =
+  QExpr (PgExpressionSyntax . mappend (emit "array_shuffle") . pgParens . fromPgExpression <$> arr)
+
+-- | Postgres array_sample(array, n) function.
+-- Randomly selects @n@ items from the array. For multidimensional arrays,
+-- an "item" is a slice with a given first subscript.
+--
+-- Precondition: @n@ must not exceed the length of the first dimension.
+arraySample_
+  :: Integral n
+  => QGenExpr ctxt Postgres s (V.Vector a)
+  -> QGenExpr ctxt Postgres s n
+  -> QGenExpr ctxt Postgres s (V.Vector a)
+arraySample_ (QExpr arr) (QExpr n) =
+  QExpr (PgExpressionSyntax . mappend (emit "array_sample") . pgParens . mconcat <$> sequenceA
+    [ fromPgExpression <$> arr
+    , pure (emit ", ")
+    , fromPgExpression <$> n
+    ])
+
+-- | Postgres array_to_string(array, delimiter) function.
+-- Converts each element to text and joins with the delimiter. NULLs are omitted.
+arrayToString_
+  :: QGenExpr ctxt Postgres s (V.Vector a)
+  -> QGenExpr ctxt Postgres s text
```

Comment:
Why is the `text` type variable not `Text`? The [PostgreSQL documentation](https://www.postgresql.org/docs/current/functions-array.html) makes it clear that the delimiter is `TEXT`, and the result is `TEXT` also. I would expect:

```haskell
arrayToString_
  :: QGenExpr ctxt Postgres s (V.Vector a)
  -> QGenExpr ctxt Postgres s Text
  -> QGenExpr ctxt Postgres s Text
```